README
1.Script "I7565H1H2_install"

  This script does not support LP-8x4x device. Please refer manual to check driver install method.

  [1]Install I7565-H1/H2 driver

     ./I7565H1H2_install"

  [2]Remove I7565-H1/H2 driver

     ./I7565H1H2_install remove"
     Remove I7565-H1/H2 driver

  [3]When conneting I7565-H1/H2 device, cdc_acm driver would insmod automatic, we send it to blacklist, you can use below command to restore.

     ./I7565H1H2_install cdc_acm"
     Remove cdc_acm from blacklist.

3.Compile demo i7565H1H2.c

  command 'make' will generate two executable files in /examples for x86 Architecture.

  command 'make LinPAC=ARM' will generate two executable files in /examples for ARM Architecture.


ChangeLog
The i-7565-H1/H2 Software for Linux
v 0.4.5 2022.3.14 by Winson
  Support ARM lib.(arm-linux-gnueabihf-gcc)

v 0.4.4 2020.3.23 by Winson
  Add boot time mount driver function in I7565H1H2_install script.
  Add udev rule to keep cdc_acm driver when you using I7565H1H2 device.

v 0.4.3 2019.2.22 by Golden
  Fix the bug of libi7k(ver 0.8.7) function "Send_Receive_Cmd" and "Receive_Cmd".
  The install script "I7565H1H2_install" can detect and install driver automatically for i-7565-H1/H2.

v 0.4.2 2017.10.11 by Winson
  Fix compile path in Makefile.

v 0.4.1 2017.4.6 by Winson
  Fix VCI_CloseCAN() thread problem.
  Rename demo "i7565H1H2.a" to "i7565H1H2_a"
  Rename demo "i7565H1H2.so" to "i7565H1H2_so"

v 0.4.0 2017.3.23 by Winson
  Fix Open CAN port fail when receive data continuously.

v 0.3.8 2016.9.30 by Winson
  Fix question that can't receive data and send data at same time.

v 0.3.7 2016.07.01 by Winson
  Add function VCI_Get_TxMsgCnt(), VCI_Clr_TxMsgBuf(), VCI_Get_TxSentCnt(), VCI_Clr_TxSentCnt()
  Modify function VCI_Get_MODInfo().
  Modify demo "i7565H1H2.c".
  Modify Manual.

v 0.3.6 2016.06.06 by Winson
  Add function VCI_OpenCAN_Ex(), VCI_Get_CANBaud_BitTime(), VCI_Set_MOD_Ex().

v 0.3.5 2015.08.24 by Winson
  Add script "I7565H1H2_install" to install I7565-H1/H2 driver.
  Modify demo i7565H1H2.c to find device file name automatic.
  Modify Manual.

v 0.3.4 2014.12.3 by Winson
  The version of "i7k" library is "0.8.4".

  fix /examples/Makefile error to support library for ARM.

v 0.3.3 2014.09.4 by Winson
  Fix /src/Makefile error to support Ubuntu.
  Fix link error when link pthread in Ubuntu.
  The command "make" will generated two execution file.
  Modify Manual.

v 0.3.2 2014.07.22 by Winson
  Fix /examples/i7565H1H2.c warning from Ubuntu.

v 0.3.1 2014.05.26 by Winson
  The version of "i7k" library is "0.7.0"

  Support to Ubuntu for kernel version 3.12.0.
  Support to Linux OS 32bit and 64bit.

v 0.3.0 2010.12.13 by Golden
  The version of "i7k" library is "0.7.0".

  Support 256 i7565H1/H2 modules most much at the same time.

v 0.2.0 2010.04.12 by Golden
  The version of "i7k" library is "0.5.2".

  Fixed the bug that I-7565-H1/H2 send command error.

v 0.1.0 2009.12.14 by Golden
  The version of "i7k" library is "0.5.1".

  Created, dirty codes.


include/msw.h
/**********************************************************************
 *
 *  msw.h
 *
 *  Header for msw.c (Microsoft Windows Warp)
 *
 *  Since 1999.10.27 by Reed Lai
 *
 **********************************************************************/
/* *INDENT-OFF* */
#ifndef _MSW_H
#define _MSW_H

#include "i7k.h"
#include "common.h"

#define DEFAULT_READ_BUF_SIZE 250

#define MAX_CMD_SIZE  250

#ifndef I7565H1H2
#define MAX_COM_PORTS 4
#define COM1          1
#define COM2          2
#define COM3          3
#define COM4          4
#else
#define MAX_COM_PORTS 256
#define USB0          0
#define USB1          1
#define USB2          2
#define USB3          3
#define USB4          4
#define USB5          5
#define USB6          6
#define USB7          7
#define USB8          8
#define USB9          9
#define USB10         10
#define USB11         11
#define USB12         12
#define USB13         13
#define USB14         14
#define USB15         15
#define USB16         16
#define USB17         17
#define USB18         18
#define USB19         19
#define USB20         20
#define USB21         21
#define USB22         22
#define USB23         23
#define USB24         24
#define USB25         25
#define USB26         26
#define USB27         27
#define USB28         28
#define USB29         29
#define USB30         30
#define USB31         31
#define USB32         32
#define USB33         33
#define USB34         34
#define USB35         35
#define USB36         36
#define USB37         37
#define USB38         38
#define USB39         39
#define USB40         40
#define USB41         41
#define USB42         42
#define USB43         43
#define USB44         44
#define USB45         45
#define USB46         46
#define USB47         47
#define USB48         48
#define USB49         49
#define USB50         50
#define USB51         51
#define USB52         52
#define USB53         53
#define USB54         54
#define USB55         55
#define USB56         56
#define USB57         57
#define USB58         58
#define USB59         59
#define USB60         60
#define USB61         61
#define USB62         62
#define USB63         63
#define USB64         64

#define USB65         65
#define USB66         66
#define USB67         67
#define USB68         68
#define USB69         69
#define USB70         70
#define USB71         71
#define USB72         72
#define USB73         73
#define USB74         74
#define USB75         75
#define USB76         76
#define USB77         77
#define USB78         78
#define USB79         79
#define USB80         80
#define USB81         81
#define USB82         82
#define USB83         83
#define USB84         84
#define USB85         85
#define USB86         86
#define USB87         87
#define USB88         88
#define USB89         89
#define USB90         90
#define USB91         91
#define USB92         92
#define USB93         93
#define USB94         94
#define USB95         95
#define USB96         96
#define USB97         97
#define USB98         98
#define USB99         99
#define USB100        100
#define USB101        101
#define USB102        102
#define USB103        103
#define USB104        104
#define USB105        105
#define USB106        106
#define USB107        107
#define USB108        108
#define USB109        109
#define USB110        110
#define USB111        111
#define USB112        112
#define USB113        113
#define USB114        114
#define USB115        115
#define USB116        116
#define USB117        117
#define USB118        118
#define USB119        119
#define USB120        120
#define USB121        121
#define USB122        122
#define USB123        123
#define USB124        124
#define USB125        125
#define USB126        126
#define USB127        127
#define USB128        128

#define USB129        129
#define USB130        130
#define USB131        131
#define USB132        132
#define USB133        133
#define USB134        134
#define USB135        135
#define USB136        136
#define USB137        137
#define USB138        138
#define USB139        139
#define USB140        140
#define USB141        141
#define USB142        142
#define USB143        143
#define USB144        144
#define USB145        145
#define USB146        146
#define USB147        147
#define USB148        148
#define USB149        149
#define USB150        150
#define USB151        151
#define USB152        152
#define USB153        153
#define USB154        154
#define USB155        155
#define USB156        156
#define USB157        157
#define USB158        158
#define USB159        159
#define USB160        160
#define USB161        161
#define USB162        162
#define USB163        163
#define USB164        164
#define USB165        165
#define USB166        166
#define USB167        167
#define USB168        168
#define USB169        169
#define USB170        170
#define USB171        171
#define USB172        172
#define USB173        173
#define USB174        174
#define USB175        175
#define USB176        176
#define USB177        177
#define USB178        178
#define USB179        179
#define USB180        180
#define USB181        181
#define USB182        182
#define USB183        183
#define USB184        184
#define USB185        185
#define USB186        186
#define USB187        187
#define USB188        188
#define USB189        189
#define USB190        190
#define USB191        191
#define USB192        192

#define USB193        193
#define USB194        194
#define USB195        195
#define USB196        196
#define USB197        197
#define USB198        198
#define USB199        199
#define USB200        200
#define USB201        201
#define USB202        202
#define USB203        203
#define USB204        204
#define USB205        205
#define USB206        206
#define USB207        207
#define USB208        208
#define USB209        209
#define USB210        210
#define USB211        211
#define USB212        212
#define USB213        213
#define USB214        214
#define USB215        215
#define USB216        216
#define USB217        217
#define USB218        218
#define USB219        219
#define USB220        220
#define USB221        221
#define USB222        222
#define USB223        223
#define USB224        224
#define USB225        225
#define USB226        226
#define USB227        227
#define USB228        228
#define USB229        229
#define USB230        230
#define USB231        231
#define USB232        232
#define USB233        233
#define USB234        234
#define USB235        235
#define USB236        236
#define USB237        237
#define USB238        238
#define USB239        239
#define USB240        240
#define USB241        241
#define USB242        242
#define USB243        243
#define USB244        244
#define USB245        245
#define USB246        246
#define USB247        247
#define USB248        248
#define USB249        249
#define USB250        250
#define USB251        251
#define USB252        252
#define USB253        253
#define USB254        254
#define USB255        255
#endif

#define EXPORTS
#define CALLBACK
#define WINAPI
#define FAR
#define NEAR

#define Data5Bit	5
#define Data6Bit	6
#define Data7Bit	7
#define Data8Bit	8

#define NonParity	0
#define OddParity	1
#define EvenParity	2

#define OneStopBit	0
#define One5StopBit	1
#define TwoStopBit	2


typedef __bool_t BOOL;

typedef u_byte_t BYTE;
typedef u_word_t WORD;
typedef u_dword_t DWORD;

typedef unsigned char UCHAR;

extern int _fd[];

extern EXPORTS WORD CALLBACK Receive_Cmd(char cPort, char szResult[],										 WORD wTimeOut, WORD wChksum);

/* Macros to warp MS-Windows functions */
#define Sleep(a)

/* Macros to warp Turbo C functions */
#define getch()   (int)getchar()
#define getche()  (int)getchar()
#define kbhit()   (int)getchar()

/* Macros to warp DOS functions */
#define RECEIVE_CMD(a, b, c, d) (int)Receive_Cmd((char)(a), (b), (long)(c), (int)(d))
#define SEND_CMD(a,b,c,d) (int)Send_Cmd((char)(a), (b), (WORD)(d))
#define OPEN_COM(p,b)     (int)Open_Com((char)(p), (DWORD)(b), Data8Bit, NonParity, OneStopBit)
#define CLOSE_COM(p)      (int)Close_Com((char)(p))

WORD Send_Receive_Cmd(char cPort, char szCmd[], char szResult[], WORD wTimeOut, WORD wChksum, WORD * wT);
//WORD Receive_Cmd(char cPort, char szResult[], WORD wTimeOut, WORD wChksum);
WORD Receive_Binary(char cPort, char szResult[], WORD wTimeOut, DWORD wLen, WORD *wT);
WORD Send_Cmd(char cPort, char szCmd[], WORD wChksum);
WORD Send_Binary(char cPort, char szCmd[],int iLen);
extern BOOL Close_Com(char port);
extern WORD Open_Com(char port, DWORD baudrate, char cData, char cParity, char cStop);


WORD SetLineStatus(BYTE cPort, BYTE DTR_Trig, BYTE RTS_Trig);
WORD FW_C2U_SendTrig(BYTE cPort, BYTE Trig);

#include "i7000.h"

/*#ifndef I7565H1H2
#include "i8000.h"
#include "i87000.h"
#endif
*/
#endif							/* _MSW_H */


include/sio.h
/*
 *      sio.h
 *
 *      v 0.00 1999.10.8 by Reed Lai
 *
 *      header of sio.c
 *
 *      History
 *
 *              0.00    create
 */

#ifndef _SIO_H
#define _SIO_H

#include <stdio.h>				/* Standard input/output definitions */
#include <stdlib.h>
#include <string.h>
#include <unistd.h>				/* UNIX standard function definitions */
#include <fcntl.h>				/* File control definitions */
#include <errno.h>				/* Error number definitions */
#include <termios.h>			/* POSIX terminal control definitions */

#include "codes.h"
#include "common.h"
#include "debug.h"

/* Error Codes */
#define OK_PORT_CLOSE   0
#define ERR_PORT_CLOSE  -1
#define ERR_PORT_OPEN   0		/* DO NOT CHANGE THIS!! */

/* sio devices */
#define SIO_DEV0     "/dev/ttyS0"
#define SIO_DEV1     "/dev/ttyS1"
#define SIO_DEV2     "/dev/ttyS2"
#define SIO_DEV3     "/dev/ttyS3"
#define SIO_DEV4     "/dev/ttyS4"

/* CAN devices */
/*
#define USB_DEV0     "/dev/ttyUSB0"
#define USB_DEV1     "/dev/ttyUSB1"
#define USB_DEV2     "/dev/ttyUSB2"
#define USB_DEV3     "/dev/ttyUSB3"
*/
/* data bits */
#define DATA_BITS_5  CS5
#define DATA_BITS_6  CS6
#define DATA_BITS_7  CS7
#define DATA_BITS_8  CS8

/* parity check */
#define NO_PARITY    0
#define ODD_PARITY   1
#define EVEN_PARITY  2

/* stop bit(s) */
#define ONE_STOP_BIT    0
#define ONE_5_STOP_BIT  1		/* but there is no 1.5 stop bit to be */
				/* defined in Linux POSIX control mode flags */
#define ONE_P5_STOP_BIT 1		/* but there is no 1.5 stop bit defined */
				/* in Linux POSIX control mode flags */
#define TWO_STOP_BIT    2
#define TWO_STOP_BITS   2

struct sio {
	struct sio *next;			/* next opened sio device */
	int fd;						/* file descriptor of this port */
	const char *name;			/* device name */
	struct termios *options;	/* current options of this port */
	struct termios *old_options;	/* original options */
};

extern void sio_set_noncan(int fd);
extern void sio_set_canon(int fd);
extern void sio_set_timer(int fd, cc_t vmin, cc_t vtime);
extern void sio_poptions(int fd);
extern int sio_set_baud(int fd, speed_t baud);
extern int sio_open(const char *port, speed_t baud, tcflag_t data,
					tcflag_t parity, tcflag_t stop);
extern int sio_close(int fd);

#endif							/* _SIO_H */


include/common.h
/**********************************************************************
 *
 *  common.h
 *
 *  v 0.0.1 2000.1.11 by Reed Lai
 *
 *  common types and definitions
 *
 *  History:
 *
 *  v 0.0.1 2000.1.11 by Reed Lai
 *      re-typedef, Linux kernel 2.2.13, Slackware 7
 *
 *  v 0.00 1999.11.03 by Reed Lai
 *      create, Linux kernel 2.2.6, Slackware 4
 *
 **********************************************************************/
#ifndef _COMMON_H
#define _COMMON_H

#include <linux/version.h>
#include <sys/types.h>

#if LINUX_VERSION_CODE == KERNEL_VERSION(2,2,6)
typedef __u8 u_byte_t;
typedef __u16 u_word_t;
typedef __u32 u_dword_t;
typedef __u64 u_qword_t;
typedef __s8 byte_t;
typedef __s16 word_t;
typedef __s32 dword_t;
typedef __s64 qword_t;
#elif LINUX_VERSION_CODE == KERNEL_VERSION(2,2,13)
typedef u_int8_t u_byte_t;
typedef u_int16_t u_word_t;
typedef u_int32_t u_dword_t;
typedef u_int64_t u_qword_t;
typedef int8_t byte_t;
typedef int16_t word_t;
typedef int32_t dword_t;
typedef int64_t qword_t;
#else
typedef unsigned char u_byte_t;
typedef unsigned short int u_word_t;
typedef unsigned int u_dword_t;
typedef unsigned long long int u_qword_t;
typedef char byte_t;
typedef short int word_t;
typedef int dword_t;
typedef long long int qword_t;
#endif

#ifndef FALSE
#  define FALSE  0
#endif

#ifndef TRUE
#  define TRUE   1
#endif

typedef int __bool_t;			/* Boolean, true or false */

typedef qword_t __tod_t;

#define MIN(a,b)  ( (a) < (b) ? (a) : (b) )
#define MAX(a,b)  ( (a) > (b) ? (a) : (b) )
#define ABS(a)    ( (a) < 0 ? -(a) : (a) )

#ifndef CTRL
#  define CTRL(x)    ((x) & 0x1f)
#endif

#endif							/* _COMMON_H */


include/global.h
#include "i7565H1H2.h"

#ifndef SUCCESS
#define SUCCESS 0
#endif

#ifndef FAILURE
#define FAILURE -1
#endif

/************    VCI_CAN Global Define   **************/
//#define SWAP(x,y)                       {DWORD t; t = x; x = y; y = t;}
//#define Word_SWAP(x,y)          {WORD t; t = x; x = y; y = t;}
#define RULE_MAX	512
#define CAN1_RecvFlag	0

/************    VCI_CAN Global Variable   **************/
DWORD CANBaud_Arr[15]=
{
	CANBaud_5K,	CANBaud_10K,	CANBaud_20K,	CANBaud_40K,
	CANBaud_50K,	CANBaud_80K,	CANBaud_100K,	CANBaud_125K,
	CANBaud_200K,	CANBaud_250K,	CANBaud_400K,	CANBaud_500K,
	CANBaud_600K,	CANBaud_800K,	CANBaud_1000K,
};

/************    VCI_CAN RxBuffer Variable   **************/
//BYTE U2C_Buf[MAX_I7565_SUPPORT][2048]={0};

//#define CAN_BUF_SIZE            65536
//#define CAN_BUF_MASK            0xffff


/* CAN1->PCUSB Define */
DWORD C12P_MsgBuf_rIdx[MAX_I7565_SUPPORT] = {0};
DWORD C12P_MsgBuf_wIdx[MAX_I7565_SUPPORT] = {0};
BYTE C12P_BufFull[MAX_I7565_SUPPORT] = {0};
#define C12P_BUF_WRIDX(index) (CAN_BUF_MASK & C12P_MsgBuf_wIdx[index])
#define C12P_BUF_RDIDX(index) (CAN_BUF_MASK & C12P_MsgBuf_rIdx[index])
#define C12P_BUF_EMPTY(index) (C12P_BUF_RDIDX(index) == C12P_BUF_WRIDX(index))
#define C12P_BUF_FULL(index)  (C12P_BUF_RDIDX(index) == (CAN_BUF_MASK & (C12P_MsgBuf_wIdx[index]+1))) //(C12P_BUF_WRIDX+1))
#define C12P_BUF_COUNT(index) (CAN_BUF_MASK & (C12P_MsgBuf_wIdx[index] - C12P_MsgBuf_rIdx[index]))
//_VCI_CAN_MSG C12P_MsgBuf[MAX_I7565_SUPPORT][CAN_BUF_SIZE];

/* CAN2->PCUSB Define */
DWORD C22P_MsgBuf_rIdx[MAX_I7565_SUPPORT] = {0};
DWORD C22P_MsgBuf_wIdx[MAX_I7565_SUPPORT] = {0};
BYTE C22P_BufFull[MAX_I7565_SUPPORT] = {0};
#define C22P_BUF_WRIDX(index) (CAN_BUF_MASK & C22P_MsgBuf_wIdx[index])
#define C22P_BUF_RDIDX(index) (CAN_BUF_MASK & C22P_MsgBuf_rIdx[index])
#define C22P_BUF_EMPTY(index) (C22P_BUF_RDIDX(index) == C22P_BUF_WRIDX(index))
#define C22P_BUF_FULL(index)  (C22P_BUF_RDIDX(index) == (CAN_BUF_MASK & (C22P_MsgBuf_wIdx[index]+1))) //(C22P_BUF_WRIDX+1))
#define C22P_BUF_COUNT(index) (CAN_BUF_MASK & (C22P_MsgBuf_wIdx[index] - C22P_MsgBuf_rIdx[index]))
//_VCI_CAN_MSG C22P_MsgBuf[MAX_I7565_SUPPORT][CAN_BUF_SIZE];

/* PCUSB->CAN1 Define */
DWORD P2C1_MsgBuf_rIdx[MAX_I7565_SUPPORT] = {0};
DWORD P2C1_MsgBuf_wIdx[MAX_I7565_SUPPORT] = {0};
BYTE P2C1_BufFull[MAX_I7565_SUPPORT] = {0};
#define P2C1_BUF_WRIDX(index) (CAN_BUF_MASK & P2C1_MsgBuf_wIdx[index])
#define P2C1_BUF_RDIDX(index) (CAN_BUF_MASK & P2C1_MsgBuf_rIdx[index])
#define P2C1_BUF_EMPTY(index) (P2C1_BUF_RDIDX(index) == P2C1_BUF_WRIDX(index))
#define P2C1_BUF_FULL(index) (P2C1_BUF_RDIDX(index) == (CAN_BUF_MASK & (P2C1_MsgBuf_wIdx[index]+1))) //(U2C1_BUF_WRIDX+1))
#define P2C1_BUF_COUNT(index) (CAN_BUF_MASK & (P2C1_MsgBuf_wIdx[index] - P2C1_MsgBuf_rIdx[index]))
//DWORD P2C1_MsgBuf[MAX_I7565_SUPPORT][CAN_BUF_SIZE][5];

/* PCUSB->CAN2 Define */
DWORD P2C2_MsgBuf_rIdx[MAX_I7565_SUPPORT] = {0};
DWORD P2C2_MsgBuf_wIdx[MAX_I7565_SUPPORT] = {0};
BYTE P2C2_BufFull[MAX_I7565_SUPPORT] = {0};
#define P2C2_BUF_WRIDX(index) (CAN_BUF_MASK & P2C2_MsgBuf_wIdx[index])
#define P2C2_BUF_RDIDX(index) (CAN_BUF_MASK & P2C2_MsgBuf_rIdx[index])
#define P2C2_BUF_EMPTY(index) (P2C2_BUF_RDIDX(index) == P2C2_BUF_WRIDX(index))
#define P2C2_BUF_FULL(index)  (P2C2_BUF_RDIDX(index) == (CAN_BUF_MASK & (P2C2_MsgBuf_wIdx[index]+1))) //(U2C2_BUF_WRIDX+1))
#define P2C2_BUF_COUNT(index) (CAN_BUF_MASK & (P2C2_MsgBuf_wIdx[index] - P2C2_MsgBuf_rIdx[index]))
//DWORD P2C2_MsgBuf[MAX_I7565_SUPPORT][CAN_BUF_SIZE][5];


/************    VCI_CAN  thread Variable   **************/
typedef struct receivethread rxthread_t;
static rxthread_t *rx_thread[MAX_I7565_SUPPORT] = {0};
static int enable_rxthread[MAX_I7565_SUPPORT] = {0};



include/i7565H1H2.h
/**********************************************************************
 *
 *  i7565H1H2.h
 *
 *  Header for i7565H1H2.c
 *
 *  Since 2009.11.09 by Golden Wang
 *
 **********************************************************************/
#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <sys/ioctl.h>
#include <unistd.h>
#include <string.h>
#include <signal.h>
#include <pthread.h>
#include <unistd.h>
#include <math.h>
#include "msw.h"

/* Common Return Code */
#define No_Err			0	//No Error
#define DEV_ModName_Err		1	//The Module Name Error
#define DEV_ModNotExist_Err	2	//The Module doesn't exist in this Port
#define DEV_PortNotOpen_Err	3	//The Port doesn't Open
#define DEV_PortClose_Err	4	//The Port doesn't Open
#define DEV_Reset_Err		5	//Reset I-7565 module failure
#define CAN_ConfigureFail_Err	6	//CAN Hardware Init Fail
#define CAN_Hardware_Err	7	//CAN Hardware Init Fail
#define CAN_PortNo_Err		8	//The Device doesn't support this CAN Port
#define CAN_FIDLength_Err	9	//The CAN Filter-ID Number exceed Max Number
#define CAN_DevDisconnect_Err	10	//The Connection of device is broken
#define CAN_TimeOut_Err		11	//The Config Command Timeout
#define CAN_ConfigCmd_Err	12	//The Config Command doesn't support
#define CAN_ConfigBusy_Err	13	//The Config Command is busy
#define CAN_RxBufEmpty		14	//The CAN Receive Buffer is empty
#define CAN_TxBufFull		15	//The CAN Send Buffer is full
#define CAN_EnableHWCyclicTxMsg_Err 16  //Enable Hardware Cyclic Tx Message Error
#define CreateRxThread_Err 17  		//Create Rx Thread Error
#define RestartRxThread_Err 18		//Restartup Rx Thread Error

/************************      VCI_CAN API Variable      *************************/ 
//Common Define
#define I7565_LIBRARY_VERSION "0.4.5"
#define MAX_I7565_SUPPORT 256
#define CAN1			1
#define CAN2			2
#define CONFIG_MODE		0
#define COMM_MODE		1
#define CAN_BUF_SIZE            65536
#define CAN_BUF_MASK            0xffff

//Device Type
#define I7565H1			1
#define I7565H2			2
	
//CAN Baudrate
#define CANBaud_5K		5000
#define CANBaud_10K		10000
#define CANBaud_20K		20000
#define CANBaud_40K		40000
#define CANBaud_50K		50000
#define CANBaud_80K		80000
#define CANBaud_100K		100000
#define CANBaud_125K		125000
#define CANBaud_200K		200000
#define CANBaud_250K		250000
#define CANBaud_400K		400000
#define CANBaud_500K		500000
#define CANBaud_600K		600000
#define CANBaud_800K		800000
#define CANBaud_1000K		1000000

/* ICPDAS VCICAN Series Message Structure Format */
//[1]. VCICAN - Module Info Format
typedef struct _VCI_MODULE_INFO{
	char	Mod_ID[12];
	char	FW_Ver[12];
	char	HW_SN[16];
} _VCI_MODULE_INFO, *PVCI_MOD_INFO; 

//[2]. VCICAN - CAN Message Format
typedef struct _VCI_CAN_MSG{
        BYTE    Mode;
        BYTE    RTR;
        BYTE    DLC;
        BYTE    Reserved;
        DWORD   ID;
        DWORD   TimeL;
        DWORD   TimeH;
        BYTE    Data[8];
} _VCI_CAN_MSG, *PVCI_CAN_MSG;


//[3]. VCICAN - CAN Parameter (CANBaud: Unit(bps))
typedef struct _VCI_CAN_PARAM{
	BYTE  PortOpen;
	BYTE  DevPort;
	BYTE  DevType;
	DWORD CAN1_Baud;
	DWORD CAN2_Baud;
	//DWORD C12P_MsgBuf_rIdx;
	//DWORD C12P_MsgBuf_wIdx;
	//BYTE  C12P_BufFull;
	_VCI_CAN_MSG C12P_MsgBuf[CAN_BUF_SIZE];
	//DWORD C22P_MsgBuf_rIdx;
        //DWORD C22P_MsgBuf_wIdx;
	//BYTE  C22P_BufFull
	_VCI_CAN_MSG C22P_MsgBuf[CAN_BUF_SIZE];
	DWORD P2C1_MsgBuf[CAN_BUF_SIZE][5];
	DWORD P2C2_MsgBuf[CAN_BUF_SIZE][5];
} _VCI_CAN_PARAM, *PVCI_CAN_PARAM;

typedef struct _VCI_CAN_PARAM_EX{
	_VCI_CAN_PARAM vc;
	BYTE	CAN1_T2Val;
	BYTE	CAN2_T2Val;
	BYTE	Reserved[32];
} _VCI_CAN_PARAM_EX, *PVCI_CAN_PARAM_EX;

//[4]. VCICAN - CAN FilterID Format
typedef struct _VCI_CAN_FilterID{
	WORD	SSFF_Num;
	WORD	GSFF_Num;
	WORD	SEFF_Num;
	WORD	GEFF_Num;
	WORD	SSFF_FID[512];
	DWORD	GSFF_FID[512];
	DWORD	SEFF_FID[512];
	DWORD	GEFF_FID[512];
} _VCI_CAN_FilterID, *PVCI_CAN_FID;

//[5]. VCICAN - CAN Status Format
typedef struct _VCI_CAN_STATUS{
	DWORD	CurCANBaud;
	BYTE	CANReg;
	BYTE	CANTxErrCnt;
	BYTE	CANRxErrCnt;
	BYTE	MODState;
	DWORD	Reserved;
} _VCI_CAN_STATUS, *PVCI_CAN_STATUS;

/************************      VCI_CAN API Function      *************************/ 
/********************     [1] Init Function      ********************/
int VCI_OpenCAN(PVCI_CAN_PARAM pCANPARAM);
int VCI_OpenCAN_NoStruct(BYTE DevPort, BYTE DevType, DWORD CAN1_Baud, DWORD CAN2_Baud);
int VCI_CloseCAN(PVCI_CAN_PARAM pCANPARAM);


/********************     [2] Module Config Function     ********************/
int VCI_Set_CANFID(BYTE DevPort, BYTE CAN_No, PVCI_CAN_FID pCANFID);
int VCI_Set_CANFID_NoStruct(BYTE DevPort, BYTE CAN_No, WORD SSFF_Num, WORD GSFF_Num, WORD SEFF_Num, WORD GEFF_Num, WORD SSFF_FID[512], DWORD GSFF_FID[512], DWORD SEFF_FID[512], DWORD GEFF_FID[512]);
//int VCI_Set_CANFID_AllPass(BYTE DevPort, BYTE CAN_No);
int VCI_Get_CANFID(BYTE DevPort, BYTE CAN_No, PVCI_CAN_FID pCANFID);
int VCI_Get_CANFID_NoStruct(BYTE DevPort, BYTE CAN_No, WORD* SSFF_Num, WORD* GSFF_Num, WORD* SEFF_Num, WORD* GEFF_Num, WORD SSFF_FID[512], DWORD GSFF_FID[512], DWORD SEFF_FID[512], DWORD GEFF_FID[512]);
int VCI_Clr_CANFID(BYTE DevPort, BYTE CAN_No);
int VCI_Get_CANStatus(BYTE DevPort, BYTE CAN_No, PVCI_CAN_STATUS pCANStatus);
int VCI_Get_CANStatus_NoStruct(BYTE DevPort, BYTE CAN_No, DWORD* CurCANBaud, BYTE* CANReg, BYTE* CANTxErrCnt, BYTE* CANRxErrCnt, BYTE* MODState);
int VCI_Clr_BufOverflowLED(BYTE DevPort, BYTE CAN_No);
int VCI_Get_MODInfo(BYTE DevPort, PVCI_MOD_INFO pMODInfo);
int VCI_Get_MODInfo_NoStruct(BYTE DevPort, char Mod_ID[12], char FW_Ver[12], char HW_SN[16]);
int VCI_Rst_MOD(PVCI_CAN_PARAM pCANPARAM);
int VCI_Set_MOD_Ex(BYTE CfgData[512]);

/********************     [3] Communication Function     ********************/
int VCI_SendCANMsg(BYTE DevPort, BYTE CAN_No, PVCI_CAN_MSG pCANMsg);
int VCI_SendCANMsg_NoStruct(BYTE DevPort, BYTE CAN_No, BYTE Mode, BYTE RTR, BYTE DLC, DWORD ID, BYTE Data[8]);
int VCI_RecvCANMsg(BYTE DevPort, BYTE CAN_No, PVCI_CAN_MSG pCANMsg);
int VCI_RecvCANMsg_NoStruct(BYTE DevPort, BYTE CAN_No, BYTE* Mode, BYTE* RTR, BYTE* DLC, DWORD* ID, DWORD* TimeL, DWORD* TimeH, BYTE Data[8]);
int VCI_EnableHWCyclicTxMsg(BYTE DevPort, BYTE CAN_No, PVCI_CAN_MSG pCANMsg, DWORD TimePeriod, DWORD TransmitTimes);
int VCI_EnableHWCyclicTxMsg_NoStruct(BYTE DevPort, BYTE CAN_No, BYTE Mode, BYTE RTR, BYTE DLC, DWORD ID, BYTE Data[8], DWORD TimePeriod, DWORD TransmitTimes);
int VCI_DisableHWCyclicTxMsg(PVCI_CAN_PARAM pCANPARAM);


/********************     [4] Software Buffer Function     ********************/
int VCI_Get_RxMsgCnt(BYTE DevPort, BYTE CAN_No, DWORD* RxMsgCnt);
int VCI_Get_RxMsgBufIsFull(BYTE DevPort, BYTE CAN_No, BYTE* Flag);
int VCI_Clr_RxMsgBuf(BYTE DevPort, BYTE CAN_No);
int VCI_Get_TxMsgCnt(BYTE DevPort, BYTE CAN_No, DWORD* TxMsgCnt);
int VCI_Clr_TxMsgBuf(BYTE DevPort, BYTE CAN_No);
int VCI_Get_TxSentCnt(BYTE DevPort, BYTE CAN_No, DWORD* TxSentCnt);
int VCI_Clr_TxSentCnt(BYTE DevPort, BYTE CAN_No);

/********************     [5] Other Function     ********************/
char * VCI_Get_Library_Version(void);


/********************     [7] Extended Function     ********************/

int VCI_OpenCAN_Ex(PVCI_CAN_PARAM_EX pCANPARAMEx);  //done
int VCI_OpenCAN_NoStruct_Ex(BYTE DevPort, BYTE DevType, DWORD CAN1_Baud, DWORD CAN2_Baud, BYTE CAN1_T2Val, BYTE CAN2_T2Val);  //done
int VCI_Get_CANBaud_BitTime(BYTE DevPort, BYTE CAN_No, BYTE* T1Val, BYTE* T2Val, BYTE* SJWVal);
